Library/Homebrew/cmd/readall.rb

```diff
@@ -20,69 +18,23 @@ def readall
         ruby_files << rb
       end
 
-      failed = false
-      workers = (0...Hardware::CPU.cores).map do
-        Thread.new do
-          begin
-            while rb = ruby_files.pop(true)
-              # As a side effect, print syntax errors/warnings to `$stderr`.
-              failed = true if syntax_errors_or_warnings?(rb)
-            end
-          rescue ThreadError # ignore empty queue error
-          end
-        end
-      end
-      workers.each(&:join)
-      Homebrew.failed = failed
+      valid_syntax = Readall.valid_ruby_syntax?(ruby_files)
+      Homebrew.failed = true unless valid_syntax
     end
 
-    formulae = []
-    alias_dirs = []
     if ARGV.named.empty?
-      formulae = Formula.files
-      alias_dirs = Tap.map(&:alias_dir)
+      if ARGV.include?("--aliases")
+        valid_aliases = Readall.valid_aliases?(Tap.map(&:alias_dir))
+        Homebrew.failed = true unless valid_aliases
+      end
+      valid_formulae = Readall.valid_formulae?(Formula.files)
+      Homebrew.failed = true unless valid_formulae
     else
       tap = Tap.fetch(ARGV.named.first)
       raise TapUnavailableError, tap.name unless tap.installed?
-      formulae = tap.formula_files
-      alias_dirs = [tap.alias_dir]
-    end
-
-    if ARGV.delete("--aliases")
-      alias_dirs.each do |alias_dir|
-        next unless alias_dir.directory?
-        Pathname.glob("#{alias_dir}/*").each do |f|
-          next unless f.symlink?
-          next if f.file?
-          onoe "Broken alias: #{f}"
-          Homebrew.failed = true
-        end
-      end
-    end
-
-    formulae.each do |file|
-      begin
-        Formulary.factory(file)
-      rescue Interrupt
-        raise
-      rescue Exception => e
-        onoe "problem in #{file}"
-        puts e
-        Homebrew.failed = true
-      end
+      options = {:aliases => ARGV.include?("--aliases")}
+      valid_tap = Readall.valid_tap?(tap, options)
+      Homebrew.failed = true unless valid_tap
```

Comment:
May I suggest to restructure this entire block a bit? Now that you've nicely abstracted away things in the `Readall` module, I think it would be clearer to iterate over taps, i.e. do something like this (untested):

```ruby
options = { :aliases => ARGV.include?("--aliases") }
taps = if ARGV.named.empty?
  Tap
else
  [Tap.fetch(ARGV.named.first)]
end
taps.each do |tap|
  Homebrew.failed = true unless Readall.valid_tap?(tap, options)
end
```